## DS 7337 - Natural Language Processing

### Author: Brandon Croom

### Homework: 1

#### 1.	Install Python (if you don’t have it already), and install NLTK.  

In [1]:
#Import the required libraries
import sys
import nltk

#print out the versions to show the installation of Python and NLTK
print("Python Version: ",sys.version)
print("NLTK Version: ", nltk.__version__)

Python Version:  3.7.6 (default, Jan  8 2020, 20:23:39) [MSC v.1916 64 bit (AMD64)]
NLTK Version:  3.4.5


#### 2. Follow the instructions in chapter 1 of Bird-Klein for implementing a “lexical diversity” scoring routine.

In [2]:
#Load in the books
from nltk.book import *

#define the lexical diversity scoring routine
def lexical_diversity(text_data):
    word_count = text_word_count(text_data)
    vocab_size = text_vocab_size(text_data)
    diversity_score = vocab_size / word_count
    return diversity_score

#define the text word count
def text_word_count(text_data):
    return len(text_data)

#define the text vocabulary size
def text_vocab_size(text_data):
    return len(set(text_data))

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


In [3]:
#test the lexical diversity function
#do the initial test with the text references in the chapter to verify functionality
print("Verification Test from Chapter:")
print("Lexical Diversity Text3: ",lexical_diversity(text3))
print("Lexical Diversity Text5: ",lexical_diversity(text5))
print()
#secondary test just to verify additional texts
print("Additional Test:")
print("Lexical Diversity Text4: ",lexical_diversity(text4))
print("Lexical Diversity Text6: ",lexical_diversity(text6))

Verification Test from Chapter:
Lexical Diversity Text3:  0.06230453042623537
Lexical Diversity Text5:  0.13477005109975562

Additional Test:
Lexical Diversity Text4:  0.06617622515804722
Lexical Diversity Text6:  0.1276595744680851


#### 3.	Go to http://www.gutenberg.org/wiki/Children%27s_Instructional_Books_(Bookshelf), and obtain three texts (of different grade levels) from the “Graded Readers” section. Report the lexical diversity score of each. Explain whether the result was surprising.

In [4]:
#Import the custom reader to read the downloaded Gutenberg books. The books downloaded were:
#The Ontario Readers: Third Book (18561.txt)
#The Ontario Readers: Fourth Book (18702.txt)
#The Ontario Readers: The High School Reader (19923.txt)
from nltk.corpus import BracketParseCorpusReader
from nltk.corpus import PlaintextCorpusReader
corpus_root = "D://Users//jbcroom1//Documents//GitHub//DS7337-NLP//HW1//corpus"
file_pattern = r".*/*.*\.txt"
#ptb = BracketParseCorpusReader(corpus_root,file_pattern)
ptb = PlaintextCorpusReader(corpus_root,file_pattern)
ptb.fileids()

['18561.txt', '18702.txt', '19923.txt']

In [5]:
print("Lexical Diversity The Ontario Readers: Third Book: ",lexical_diversity(ptb.words(ptb.fileids()[0])))
print("Lexical Diversity The Ontario Readers: Fourth Book: ",lexical_diversity(ptb.words(ptb.fileids()[1])))
print("Lexical Diversity The Ontario Readers: The High School Book: ",lexical_diversity(ptb.words(ptb.fileids()[2])))

Lexical Diversity The Ontario Readers: Third Book:  0.11501215001376602
Lexical Diversity The Ontario Readers: Fourth Book:  0.12121775025799794
Lexical Diversity The Ontario Readers: The High School Book:  0.1039680780314786


The lexical diversity of these books was suprising. We see that the book designed for high school has less lexical diversity than the books designed for younger readers. It's diversity is 10.4%. The book designed for fourth year has the highest diversity at 12.12%, while the book designed for third year has a diversity of 11.5%. This goes against the expectation that as a student increases in grade level the reading diversity should increase. In this particular instance we see a fourth year is exposed to greater lexical diversity.

#### 4. Also compare the vocabulary size of the same three texts. Explain whether the result was surprising.  

In [6]:
print("Vocabulary Size: The Ontario Readers: Third Book: ",text_vocab_size(ptb.words(ptb.fileids()[0])))
print("Word Count: The Ontario Readers: Third Book: ",text_word_count(ptb.words(ptb.fileids()[0])))

print("Vocabulary Size: The Ontario Readers: Fourth Book: ",text_vocab_size(ptb.words(ptb.fileids()[1])))
print("Word Count: The Ontario Readers: Fourth Book: ",text_word_count(ptb.words(ptb.fileids()[1])))

print("Vocabulary Size: The Ontario Readers: The High School Book: ",text_vocab_size(ptb.words(ptb.fileids()[2])))
print("Word Count: The Ontario Readers: The High School Book: ",text_word_count(ptb.words(ptb.fileids()[2])))

Vocabulary Size: The Ontario Readers: Third Book:  9608
Word Count: The Ontario Readers: Third Book:  83539
Vocabulary Size: The Ontario Readers: Fourth Book:  11746
Word Count: The Ontario Readers: Fourth Book:  96900
Vocabulary Size: The Ontario Readers: The High School Book:  16415
Word Count: The Ontario Readers: The High School Book:  157885


#### 5. Write a paragraph arguing whether vocabulary size and lexical diversity in combination could be a better measure of text difficulty (or reading level) than either measure is by itself.

In looking at the data provided it makes sense to combine the vocabulary size and lexical diversity as a measure of text difficulty. Using the examples above as a baseline, we've already shown that just looking at lexical diversity has interesting results. We see from those initial comparisons that the book designed for high school has a lower lexical diversity when compared to the other books. These resuls are a bit counter intuitive. However, looking at the vocabulary size we see that the high school book has both a larger word count and a larger vocabulary size when compared to the other two books. This logically maskes sense as we would expect a high school based book to have a larger range of vocabulary. Taking both the lexical diversity and the vocabulary together is a much more adequate reference of text difficulty. We can have a book that has lower lexical diversity but higher vocabulary size. 